borsh-rs/borsh/src/de/mod.rs

```diff
@@ -22,25 +24,98 @@ pub trait BorshDeserialize: Sized {
         }
         Ok(result)
     }
+
+    /// Returns the size in bytes of Self, if the size is fixed without dynamic memory inside,
+    /// e.g. u32 -> Some(4)
+    fn get_fixed_size() -> Option<usize> {
+        None
+    }
+
+    /// Whether Self is u8.
+    /// NOTE: `Vec<u8>` is the most common use-case for serialization and deserialization, it's
+    /// worth handling it as a special case to improve performance.
+    /// It's a workaround for specific `Vec<u8>` implementation versus generic `Vec<T>`
+    /// implementation. See https://github.com/rust-lang/rfcs/pull/1210 for details.
+    #[inline]
+    fn is_u8() -> bool {
+        false
+    }
 }
 
 impl BorshDeserialize for u8 {
     #[inline]
-    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error> {
-        let mut res = 0u8;
-        reader.read_exact(std::slice::from_mut(&mut res))?;
+    fn deserialize<R: BufRead>(reader: &mut R) -> Result<Self, Error> {
+        let res = {
+            let buf = reader.fill_buf()?;
+            if buf.is_empty() {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+                ));
+            }
+            buf[0]
+        };
+        reader.consume(size_of::<u8>());
         Ok(res)
     }
+
+    #[inline]
+    fn try_from_slice(v: &[u8]) -> Result<Self, Error> {
```

Comment:
We don't really need to override it here. Plus this function has a different behavior than the original that throws when not all bytes were read.